contracts/src/NodeRegistry.sol

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+
+contract NodeRegistry {
+    /**
+     * STATE *
+     */
+    /// Node publicKey => Node struct
+    mapping(string => Node) public whitelist;
+
+    string public linkedListHead;
+    uint256 public whitelistCount;
+    bool private initialized;
+
+    struct Node {
+        address owner;
+        string primaryAddress;
+        string networkKey;
+        Worker[] workers;
+        string previous;
+        string next;
+    }
+
+    struct NodeInfo {
+        address owner;
+        string primaryPublicKey;
+        string primaryAddress;
+        string networkKey;
+        Worker[] workers;
+    }
+
+    struct Worker {
+        string workerAddress;
+        string workerPublicKey;
+        string workerMempool;
+    }
+
+    function initialize(NodeInfo[] memory _genesis_committee) external {
+        require(!initialized, "contract already initialized");
+        for (uint256 i; i < _genesis_committee.length;) {
+            _registerNode(_genesis_committee[i]);
+            unchecked {
+                i += 1;
+            }
+        }
+        initialized = true;
+    }
+
+    function registerNode(NodeInfo memory _node) external {
+        require(whitelist[_node.primaryPublicKey].owner == address(0), "This node is already on whitelist");
+        _registerNode(_node);
+    }
+
+    function _registerNode(NodeInfo memory _node) private {
+        whitelist[linkedListHead].previous = _node.primaryAddress;
+        string memory next = linkedListHead;
+        Node storage node = whitelist[_node.primaryPublicKey];
+        node.owner = _node.owner;
+        node.primaryAddress = _node.primaryAddress;
+        node.networkKey = _node.networkKey;
+        node.next = next;
+
+        for(uint256 i; i < _node.workers.length;){
+            node.workers.push(_node.workers[i]);
+            unchecked {i+= 1;}
+        }
+        
+        whitelistCount += 1;
+        linkedListHead = _node.primaryPublicKey;
+    }
+
+    function removeNode(string calldata _nodeAddress) external {
+        _removeNode(_nodeAddress);
+    }
+
+    function _removeNode(string calldata _nodePublicKey) private {
+        whitelist[whitelist[_nodePublicKey].previous] = whitelist[whitelist[_nodePublicKey].next];
+        whitelist[whitelist[_nodePublicKey].next] = whitelist[whitelist[_nodePublicKey].previous];
```

Comment:
```suggestion
        whitelist[whitelist[_nodePublicKey].previous].next = whitelist[whitelist[_nodePublicKey].next].primaryAddress;
        whitelist[whitelist[_nodePublicKey].next].previous = whitelist[whitelist[_nodePublicKey].previous].primaryAddress;
```
I'm probably missing something obvious here, but why aren't we doing this?
Setting the next pointer of node_to_delete.previous to node_to_delete.next and also setting the previous pointer of node_to_delete.next to node_to_delete.previous.
Also, is it necessary to check for edge cases like the list is empty or only contains one node?

Reply:
Good catch this logic was wrong i fixed. Edge cases should be covered since every value is default 0 address in solidity but we need tests on this, which will be in the next PR that gets these contracts production ready.